sdk/python/sawtooth_sdk/processor/core.py

```diff
@@ -33,65 +43,175 @@
                 self._stream.send_back(
                     message_type=Message.TP_PROCESS_RESPONSE,
                     correlation_id=msg.correlation_id,
                     content=TpProcessResponse(
                         status=TpProcessResponse.INVALID_TRANSACTION
                     ).SerializeToString())
-            except InternalError as ie:
-                LOGGER.warning("State Error! %s", ie)
+            except ValidatorConnectionError as vce:
+                LOGGER.warning("%s", vce)
+        except InternalError as ie:
+            LOGGER.warning("State Error! %s", ie)
+            try:
                 self._stream.send_back(
                     message_type=Message.TP_PROCESS_RESPONSE,
                     correlation_id=msg.correlation_id,
                     content=TpProcessResponse(
                         status=TpProcessResponse.INTERNAL_ERROR
                     ).SerializeToString())
+            except ValidatorConnectionError as vce:
+                LOGGER.warning("%s", vce)
+        except ValidatorConnectionError as vce:
+            LOGGER.warning("%s", vce)
+
+    def _process_future(self, future, timeout=None, sigint=False):
+        try:
+            msg = future.result(timeout)
+        except CancelledError:
+            return
+        if msg is RECONNECT_EVENT:
+            if sigint is False:
+                LOGGER.debug("reregistering with validator")
+                self._register()
+        else:
+            LOGGER.debug(
+                'received message of type: %s',
+                Message.MessageType.Name(msg.message_type))
+            self._process(msg)
+
+    def _register(self):
+        futures = []
+        for message in self._register_requests():
+            self._stream.wait_for_ready()
+            future = self._stream.send(
+                message_type=Message.TP_REGISTER_REQUEST,
+                content=message.SerializeToString())
+            futures.append(future)
+
+        for future in futures:
+            resp = TpRegisterResponse()
+            resp.ParseFromString(future.result().content)
+            LOGGER.debug("register attempt: %s",
+                         TpRegisterResponse.Status.Name(resp.status))
+
+    def _unregister(self):
+        futures = []
+        for message in self._unregister_requests():
+            self._stream.wait_for_ready()
+            future = self._stream.send(
+                message_type=Message.TP_UNREGISTER_REQUEST,
+                content=message.SerializeToString())
+            futures.append(future)
+        for f in futures:
+            response = TpUnRegisterResponse()
+            response.ParseFromString(f.result(1).content)
+            LOGGER.debug("unregister attempt: %s",
+                         TpUnRegisterResponse.Status.Name(response.status))
+
+    def start(self):
+        fut = None
+        try:
+            self._register()
+            while True:
```

Comment:
There probably could be a control-flag here, to be set during the shutdown phase.